### DISCUSSION TITLE
What is the proper way to figure out the address of `TxIn`?

### DISCUSSION THREAD

**Discussion by gd87429** [rust-bitcoin/rust-bitcoin#1383] [Q&A] [answered]
I'm having a hard time figuring out how to get a list of "in" addresses that went into a transaction.

If I have a [TxIn](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.TxIn.html), am I supposed to do a `get_raw_transaction_info()` for each `TxIn.previous_output.txid + vout`, and then parse addresses by trying to navigate my way around [GetRawTransactionResultVoutScriptPubKey](https://docs.rs/bitcoincore-rpc-json/0.16.0/bitcoincore_rpc_json/struct.GetRawTransactionResultVoutScriptPubKey.html), or is there a simpler way?

P.S: If there is no simpler way, how do I know which of the addresses in `pub addresses: Option<Vec<Address>>` has been used in the particular transaction?

**Comment by dpc** [comment, marked as answer]
If you have a `-txindex` on your node, you can request raw transaction by `OutPoint::txid`, decode as `Transaction` and refer to the `Output` by the index `vout`.

**Reply by gd87429** [reply]
Aha, I see, thank you. Looks like my node needs some time ("transactions are still in the process of being indexed") so I'm not able to quickly confirm I can get the code to work in the way you suggest, but I think I understand.

Is this the simplest way to go about it? I'm wondering what the alternatives look like without turning on the `-txindex` option?

Edit: Did some more reading up. Looks like `-txindex` is [the right way to go](https://bitcoin.stackexchange.com/a/35708) and `get_raw_transaction` is the function I need. @dpc Thank you for pointing me in the right direction.

**Reply by dpc** [reply]
Unless you use your own indexer (like Electrum (electrs), etc.) `-txindex` is the least involved method. Check out projects like https://github.com/RCasatta/blocks_iterator too.

**Reply by dpc** [reply]
"Bitcoin indexing" is the typical name used for the process of making the blockchain data easy and efficient to query. A node without `-txindex` maintains only the index of unspent outputs (UTXO set) so it's not suitable for querying past txes or already spent outputs.

`-txindex` is a bare minimum, but has a huge overhead, because you need to decode whole tx, just to look up one output.

**Reply by Kixunil** [reply]
If you know the block hash or height of the transaction you don't need txindex. Also if you have PSBT we have helper method for that.

**Reply by gd87429** [reply]
@Kixunil But the [TxIn](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.TxIn.html)::[OutPoint](https://docs.rs/bitcoin/latest/bitcoin/blockdata/transaction/struct.OutPoint.html) does not provide block hash or height, right?

**Reply by dpc** [reply]
Nope. You'd need an indexer of some kind for that again. AFAIK, `electrs` uses this method (mapping outpoints to height) https://github.com/romanz/electrs/blob/master/doc/schema.md#transaction-outputs-index-funding as it's leads to a very compact database. The downside is that to get your transaction, you basically need to fetch and decode the whole block from the node.

**Reply by gd87429** [reply]
Nice. Thanks for the heads up, this is all very useful 👍
